Arm/State.lean

```diff
@@ -662,3 +662,296 @@ theorem write_mem_irrelevant :
+
+def read_mem_bytes' (n : Nat) (addr : BitVec 64) (s : Memory) : BitVec (n * 8) :=
+  match n with
+  | 0 => 0#0
+  | n' + 1 =>
+    let byte := read_mem' addr s
+    let rest := read_mem_bytes' n' (addr + 1#64) s
+    have h: n' * 8 + 8 = (n' + 1) * 8 := by simp_arith
+    BitVec.cast h (rest ++ byte)
+
+theorem read_mem_bytes_eq_read_mem_bytes' (s : ArmState) :
+    read_mem_bytes n addr s = read_mem_bytes' n addr s.mem := by
+  induction n generalizing addr s
+  case zero => simp [read_mem_bytes, read_mem_bytes']
+  case succ n' ih =>
+    simp [read_mem_bytes, read_mem_bytes', read_mem_eq_read_mem', ih]
+
+@[simp]
+theorem read_mem_bytes'_zero_eq : read_mem_bytes' 0 addr s = 0#0 := rfl
+
+theorem read_mem_bytes'_succ_eq :
+  read_mem_bytes' (n' + 1) addr s = ((read_mem_bytes' n' (addr + 1) s) ++ read_mem' addr s).cast (by omega) := rfl
+
+theorem getLsb_read_mem_bytes' {n i : Nat} {addr : BitVec 64} {s : Memory} (hn : n ≤ 2^64) :
+    (read_mem_bytes' n addr s).getLsb i =
+    (decide (i < n * 8) && (s (addr + BitVec.ofNat 64 (i / 8))).getLsb (i % 8)) := by
+  induction n generalizing i addr s
+  case zero =>
+    simp
+  case succ n' ih =>
+    simp only [read_mem_bytes'_succ_eq, BitVec.ofNat_eq_ofNat, BitVec.getLsb_cast,
+      BitVec.getLsb_append, getLsb_read_mem']
+    rw [Nat.succ_mul]
+    by_cases h₁ : (i < 8)
+    · simp only [h₁, decide_True, cond_true, show i < n' * 8 + 8 by omega, Bool.true_and]
+      have hdiv : i / 8 = 0 :=  Nat.div_eq_of_lt h₁
+      rw [hdiv]
+      simp only [BitVec.add_zero]
+      have hmod : i % 8 = i := Nat.mod_eq_of_lt h₁
+      simp [hmod]
+    · simp only [h₁, decide_False, cond_false]
+      rw [ih]
+      by_cases h₂ : i - 8 < n' * 8
+      · simp only [h₂, decide_True, Bool.true_and, show (i < n' * 8 + 8) by omega]
+        have hi' : ∃ i', i = i' + 8 := by
+          apply Classical.byContradiction
+          intros h
+          simp only [not_exists] at h
+          specialize h (i - 8)
+          omega
+        obtain ⟨i', hi'⟩ := hi'
+        subst hi'
+        simp only [Nat.add_sub_cancel, Nat.zero_lt_succ, Nat.add_div_right, Nat.add_mod_right]
+        congr 2
+        rw [BitVec.add_assoc]
+        congr
+        rw [BitVec.add_def]
+        congr 1
+        simp only [BitVec.toNat_ofNat, Nat.reducePow, Nat.reduceMod]
+        rw [Nat.mod_eq_of_lt]
+        · omega
+        · omega
+      · simp only [h₂, decide_False, Bool.false_and, Bool.false_eq, Bool.and_eq_false_imp,
+          decide_eq_true_eq]
+        intros h₃
+        omega
+      · omega
+
+/--
+A variant of `write_mem` that directly talks about writes to memory, instead of over the entire `ArmState`,
+-/
+def write_mem' (addr : BitVec 64) (val : BitVec 8) (m : Memory) : Memory :=
+  write_store addr val m
+
+theorem write_mem'_of_eq (hix : ix = addr) : write_mem' addr val s ix = val := by
+  simp only [write_mem']
+  subst ix
+  apply store_read_over_write_same
```

Comment:
Should this not be phrased in terms of `read_mem'`, instead of the direct function application?

The fact that memory is implemented as a function seems like an implementation detail we shouldn't expose in theorems

Reply:
Idem for all the other theorems that talk about reading from memory in terms of the function application

Reply:
I'll mark these as `protected` in order to convey the design principle that these are meant to be low-level APIs that builds the higher level API on top of it.

Reply:
Can't mark them protected since they're in the root namespace. Rather, I'm going to write a docstring calling these out as low-level